compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala

```diff
@@ -214,38 +216,49 @@ object TypeTestsCasts {
                 testCls.is(Final) || !testCls.is(Trait) && !foundCls.is(Trait)
               )
               if (foundCls.is(Final))
-                unreachable(i"$foundCls is not a subclass of $testCls")
+                unreachable(i"type ${expr.tpe.widen} is not a subclass of $testCls")
               else if (unrelated)
-                unreachable(i"$foundCls and $testCls are unrelated")
+                unreachable(i"type ${expr.tpe.widen} and $testCls are unrelated")
               else true
             }
             else true
 
           if (expr.tpe <:< testType)
             if (expr.tpe.isNotNull) {
-              if (!inMatch) ctx.warning(TypeTestAlwaysSucceeds(foundCls, testCls), tree.sourcePos)
+              if (!inMatch) ctx.warning(TypeTestAlwaysSucceeds(expr.tpe, testType), tree.sourcePos)
               constant(expr, Literal(Constant(true)))
             }
             else expr.testNotNull
-          else if (!checkSensical)
-            constant(expr, Literal(Constant(false)))
-          else if (testCls.isPrimitiveValueClass)
-            if (foundCls.isPrimitiveValueClass)
-              constant(expr, Literal(Constant(foundCls == testCls)))
+          else {
+            val nestedCtx = ctx.fresh.setNewTyperState()
+            val foundClsSyms = foundClasses(expr.tpe.widen, Nil)
+            val sensical = foundClsSyms.exists(sym => checkSensical(sym)(using nestedCtx))
+            if (!sensical) {
+              nestedCtx.typerState.commit()
+              constant(expr, Literal(Constant(false)))
+            }
+            else if (testCls.isPrimitiveValueClass )
+              if (foundClsSyms.size == 1 && foundClsSyms.head.isPrimitiveValueClass)
+                constant(expr, Literal(Constant(foundClsSyms.head == testCls)))
+              else
+                transformIsInstanceOf(expr, defn.boxedType(testCls.typeRef), flagUnrelated)
```

Comment:
I think this would be nicer as a pattern match `foundClsSyms match { case List(foundCls) => ... case _ => ... }`